File: libs/src/csharp/SalmonWin/Sequencer/WinClientSequencer.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

namespace Salmon.Win.Sequencer;

using Mku.Salmon.Sequence;
using Mku.Sequence;
using System;
using System.IO;
using System.IO.Pipes;
using System.Security.Principal;
using System.Text;
using System.Threading;
using System.Xml;

/// <summary>
/// Sequencer client that connects to the windows service.
/// </summary>
public class WinClientSequencer : INonceSequencer
{
    private string pipeName;
    private NamedPipeClientStream client;
    private StreamReader reader;
    private StreamWriter writer;
    private bool localSystemOnly = true;

    /// <summary>
    /// The request type
    /// </summary>
    public enum RequestType
    {
        /// <summary>
        /// Create a sequence
        /// </summary>
        CreateSequence,
        /// <summary>
        /// Initialize a sequence
        /// </summary>
        InitSequence,
        /// <summary>
        /// Set the max nonce
        /// </summary>
        SetMaxNonce,
        /// <summary>
        /// Get the next nonce
        /// </summary>
        NextNonce,
        /// <summary>
        /// Revoke the current sequence
        /// </summary>
        RevokeSequence,
        /// <summary>
        ///  Get the current sequence
        /// </summary>
        GetSequence
    }

    /// <summary>
    /// Instantiate a sequencer client that connects to the Salmon Windows Service.
    /// </summary>
    /// <param name="pipeName"></param>
    public WinClientSequencer(string pipeName, bool localSystemOnly = true)
    {
        this.pipeName = pipeName;
        this.localSystemOnly = localSystemOnly;
        CreatePipeClient();
    }

    /// <summary>
    /// Create the named pipe client
    /// </summary>
    /// <exception cref="SalmonSequenceException"></exception>
    private void CreatePipeClient()
    {
        if (localSystemOnly && !IsServiceAdmin(pipeName))
        {
            throw new SalmonSequenceException("Service should run as LocalSystem account");
        }
        client = new NamedPipeClientStream(pipeName);
        client.Connect(5000);
        reader = new StreamReader(client);
        writer = new StreamWriter(client);
    }

    /// <summary>
    /// True if the service is started as Admin. This can be called to verify that 
    /// the named piped is not faux.
    /// </summary>
    /// <param name="pipeName"></param>
    /// <returns></returns>
    public static bool IsServiceAdmin(string pipeName)
    {
        NamedPipeClientStream tempClient = new NamedPipeClientStream(pipeName);
        tempClient.Connect(5000);
        PipeSecurity ac = tempClient.GetAccessControl();
        IdentityReference sid = ac.GetOwner(typeof(SecurityIdentifier));
        NTAccount acct = (NTAccount)sid.Translate(typeof(NTAccount));
        //Console.WriteLine("server owner sid: " + sid.Value);
        //Console.WriteLine("server owner name: " + acct.Value);
        var adminSid = new SecurityIdentifier(WellKnownSidType.BuiltinAdministratorsSid, null);
        //FIXME: wait till the pipe connection is released
        tempClient.Dispose();
        Thread.Sleep(1000);
        return sid == adminSid;
    }

    /// <summary>
    /// Send a request to create the sequence.
    /// </summary>
    /// <param name="driveId"></param>
    /// <param name="authId"></param>
    /// <exception cref="SalmonSequenceException"></exception>
    public void CreateSequence(string driveId, string authId)
    {
        Response res;
        try
        {
            string request = GenerateRequest(driveId, authId, RequestType.CreateSequence);
            client.WaitForPipeDrain();
            writer.WriteLine(request);
            writer.Flush();
            string response = reader.ReadLine();
            res = Response.Parse(response);
        }
        catch (Exception e)
        {
            throw new SalmonSequenceException("Could not create sequence: ", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SalmonSequenceException("Could not create sequence: " + res.error);
    }

    /// <summary>
    /// Send a request to get the current sequence
    /// </summary>
    /// <param name="driveId"></param>
    /// <returns></returns>
    /// <exception cref="SalmonSequenceException"></exception>
    public NonceSequence GetSequence(string driveId)
    {
        Response res;
        try
        {
            string request = GenerateRequest(driveId, null, RequestType.GetSequence);
            client.WaitForPipeDrain();
            writer.WriteLine(request);
            writer.Flush();
            string response = reader.ReadLine();
            res = Response.Parse(response);
        }
        catch (Exception e)
        {
            throw new SalmonSequenceException("Could not get sequence", e);
        }
        if (res.status == Response.ResponseStatus.Error)
            throw new SalmonSequenceException("Could not get sequence: " + res.error);
        if (res.status == Response.ResponseStatus.NotFound)
            return null;
        return new NonceSequence(res.driveId, res.authId,
            res.nextNonce, res.maxNonce, res.seqStatus);
    }

    /// <summary>
    /// Send a request to initialize the sequence, can only be run once.
    /// </summary>
    /// <param name="driveId"></param>
    /// <param name="authId"></param>
    /// <param name="startNonce"></param>
    /// <param name="maxNonce"></param>
    /// <exception cref="SalmonSequenceException"></exception>
    public void InitSequence(string driveId, string authId, byte[] startNonce, byte[] maxNonce)
    {
        Response res;
        try
        {
            string request = GenerateRequest(driveId, authId, RequestType.InitSequence,
                startNonce, maxNonce);
            client.WaitForPipeDrain();
            writer.WriteLine(request);
            writer.Flush();
            string response = reader.ReadLine();
            res = Response.Parse(response);
        }
        catch (Exception e)
        {
            throw new SalmonSequenceException("Could not init sequence", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SalmonSequenceException("Could not init sequence: " + res.error);
    }

    /// <summary>
    /// Send a request to get the next nonce.
    /// </summary>
    /// <param name="driveId"></param>
    /// <returns></returns>
    /// <exception cref="SalmonSequenceException"></exception>
    public byte[] NextNonce(string driveId)
    {
        Response res;
        try
        {
            string request = GenerateRequest(driveId, null, RequestType.NextNonce);
            client.WaitForPipeDrain();
            writer.WriteLine(request);
            writer.Flush();
            string response = reader.ReadLine();
            res = Response.Parse(response);
        }
        catch (Exception e)
        {
            throw new SalmonSequenceException("Could not get next nonce", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SalmonSequenceException("Could not get next nonce: " + res.error);
        return res.nextNonce;
    }

    /// <summary>
    /// Send a request to revoke the current sequence
    /// </summary>
    /// <param name="driveId"></param>
    /// <exception cref="SalmonSequenceException"></exception>
    public void RevokeSequence(string driveId)
    {
        Response res;
        try
        {
            string request = GenerateRequest(driveId, null, RequestType.RevokeSequence);
            client.WaitForPipeDrain();
            writer.WriteLine(request);
            writer.Flush();
            string response = reader.ReadLine();
            res = Response.Parse(response);
        }
        catch (Exception e)
        {
            throw new SalmonSequenceException("Could not revoke Sequence", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SalmonSequenceException("Could not revoke Sequence: " + res.error);
    }

    /// <summary>
    /// Send a request to set the max nonce.
    /// </summary>
    /// <param name="driveId"></param>
    /// <param name="authId"></param>
    /// <param name="maxNonce"></param>
    /// <exception cref="SalmonSequenceException"></exception>
    public void SetMaxNonce(string driveId, string authId, byte[] maxNonce)
    {
        string request = GenerateRequest(driveId, authId, RequestType.SetMaxNonce,
            maxNonce: maxNonce);
        client.WaitForPipeDrain();
        writer.WriteLine(request);
        writer.Flush();
        string response = reader.ReadLine();
        Response res;
        try
        {
            res = Response.Parse(response);
        }
        catch (Exception e)
        {
            throw new SalmonSequenceException("Could not revoke Sequence: " + e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SalmonSequenceException("Could not revoke Sequence: " + res.error);
    }

    /// <summary>
    /// Send a request to generate the sequence
    /// </summary>
    /// <param name="driveId"></param>
    /// <param name="authId"></param>
    /// <param name="type"></param>
    /// <param name="nextNonce"></param>
    /// <param name="maxNonce"></param>
    /// <returns></returns>
    public string GenerateRequest(string driveId, string authId, RequestType type,
            byte[] nextNonce = null, byte[] maxNonce = null)
    {
        MemoryStream stream = new MemoryStream();
        XmlWriter writer = null;
        try
        {
            XmlWriterSettings settings = new XmlWriterSettings();
            writer = XmlWriter.Create(stream, settings);
            writer.WriteStartDocument();
            writer.WriteStartElement("drive");
            writer.WriteAttributeString("driveId", driveId);
            if (authId != null)
                writer.WriteAttributeString("authId", authId);
            writer.WriteAttributeString("type", type.ToString());
            if (nextNonce != null)
                writer.WriteAttributeString("nextNonce", Convert.ToBase64String(nextNonce));
            if (maxNonce != null)
                writer.WriteAttributeString("maxNonce", Convert.ToBase64String(maxNonce));
            writer.WriteEndElement();
            writer.WriteEndDocument();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
            throw ex;
        }
        finally
        {
            if (writer != null)
            {
                try
                {
                    writer.Flush();
                    writer.Close();
                }
                catch (Exception e)
                {
                    throw new IOException("Could not close XML stream", e);
                }
            }
            stream.Flush();
            stream.Close();
        }
        return Encoding.UTF8.GetString(stream.ToArray());
    }

    /// <summary>
    /// Close the sequencer
    /// </summary>
    public void Close()
    {
        if (client != null)
        {
            try
            {
                client.Close();
            }
            catch (Exception) { }
        }
    }

    /// <summary>
    /// Responce received from the Windows service
    /// </summary>
    private class Response
    {
        internal string driveId;
        internal string authId;
        internal ResponseStatus status;
        internal NonceSequence.Status seqStatus;
        internal byte[] nextNonce;
        internal byte[] maxNonce;
        internal string error;

        /// <summary>
        /// Response status from the Windows Service
        /// </summary>
        public enum ResponseStatus
        {
            Ok, NotFound, Error
        }

        /// <summary>
        /// Parse the response
        /// </summary>
        /// <param name="contents"></param>
        /// <returns></returns>
        internal static Response Parse(string contents)
        {
            Response request = new Response();
            MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes(contents));
            try
            {
                stream.Position = 0;
                XmlDocument document = new XmlDocument();
                document.Load(stream);
                XmlNode drive = document.SelectSingleNode("/drive");
                request.driveId = drive.Attributes.GetNamedItem("driveId").Value;
                request.authId = drive.Attributes.GetNamedItem("authId").Value;
                request.status = (ResponseStatus)Enum.Parse(typeof(ResponseStatus), drive.Attributes.GetNamedItem("status").Value);
                if (drive.Attributes.GetNamedItem("seqStatus") != null)
                    request.seqStatus = (NonceSequence.Status)Enum.Parse(typeof(NonceSequence.Status), drive.Attributes.GetNamedItem("seqStatus").Value);
                if (drive.Attributes.GetNamedItem("nextNonce") != null)
                    request.nextNonce = Convert.FromBase64String(drive.Attributes.GetNamedItem("nextNonce").Value);
                if (drive.Attributes.GetNamedItem("maxNonce") != null)
                    request.maxNonce = Convert.FromBase64String(drive.Attributes.GetNamedItem("maxNonce").Value);
                if (drive.Attributes.GetNamedItem("error") != null)
                    request.error = drive.Attributes.GetNamedItem("error").Value;
            }
            catch (Exception ex)
            {
                throw ex;
            }
            finally
            {
                if (stream != null)
                {
                    stream.Close();
                }
            }
            return request;
        }
    }
}

File: libs/src/java/salmon-win/com/mku/win/salmon/sequencer/WinClientSequencer.java
package com.mku.win.salmon.sequencer;
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import com.mku.salmon.sequence.INonceSequencer;
import com.mku.salmon.sequence.SequenceException;
import com.mku.salmon.sequence.NonceSequence;
import com.sun.jna.platform.win32.AccCtrl;
import com.sun.jna.platform.win32.Advapi32;
import com.sun.jna.platform.win32.Kernel32;
import com.sun.jna.platform.win32.WinNT;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.stream.XMLOutputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamWriter;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * Nonce sequencer for use with the salmon windows service.
 */
public class WinClientSequencer implements INonceSequencer, Closeable {
    private String pipeName;
    private RandomAccessFile namedPipe;

    /**
     * Request type.
     */
    public enum RequestType {
        /**
         * Create a sequence
         */
        CreateSequence,
        /**
         * Initialize a sequence
         */
        InitSequence,
        /**
         * Set the maximum nonce of this sequence
         */
        SetMaxNonce,
        /**
         * Get the next nonce
         */
        NextNonce,
        /**
         * Revoke this sequence, any subsequent operation will fail
         */
        RevokeSequence,
        /**
         * Get a sequence
         */
        GetSequence
    }

    /**
     * Instanticate a sequencer.
     *
     * @param pipeName The name of the pipe to use see Salmon WinService.
     * @throws Exception If there is a problem with accessing the pipe.
     */
    public WinClientSequencer(String pipeName) throws Exception {
        this.pipeName = pipeName;
        createPipeClient();
    }

    private void createPipeClient() throws Exception {
        //FIXME: wait till the pipe connection is released
        Thread.sleep(1000);
        if (!isServiceAdmin(pipeName)) {
            throw new Exception("Service should run as LocalSystem account");
        }
        namedPipe = new RandomAccessFile("\\\\.\\pipe\\" + pipeName, "rw");
    }

    /**
     * Check if this is a service running as local admin.
     *
     * @param pipeName The pipe name.
     * @return True if running as local admin.
     */
    public static boolean isServiceAdmin(String pipeName) {
        WinNT.HANDLE hNamedPipe = Kernel32.INSTANCE.CreateFile(
                "\\\\.\\pipe\\" + pipeName,
                Kernel32.GENERIC_READ | Kernel32.GENERIC_WRITE,
                0,
                null,
                Kernel32.OPEN_EXISTING,
                0,
                null);

        PointerByReference ppsidOwner = new PointerByReference();
        PointerByReference ppsidGroup = new PointerByReference();
        PointerByReference ppDacl = new PointerByReference();
        PointerByReference ppSecurityDescriptor = new PointerByReference();
        Advapi32.INSTANCE.GetSecurityInfo(
                hNamedPipe,
                AccCtrl.SE_OBJECT_TYPE.SE_KERNEL_OBJECT,
                Kernel32.OWNER_SECURITY_INFORMATION
                        | Kernel32.GROUP_SECURITY_INFORMATION
                        | Kernel32.DACL_SECURITY_INFORMATION,
                ppsidOwner,
                ppsidGroup,
                ppDacl,
                null,
                ppSecurityDescriptor);
        Kernel32.INSTANCE.CloseHandle(hNamedPipe);
        WinNT.PSID ppSid = new WinNT.PSID(ppsidOwner.getValue());

        WinNT.PSID pSid = new WinNT.PSID(WinNT.SECURITY_MAX_SID_SIZE);
        IntByReference cbSid = new IntByReference(WinNT.SECURITY_MAX_SID_SIZE);
        Advapi32.INSTANCE.CreateWellKnownSid(WinNT.WELL_KNOWN_SID_TYPE.WinBuiltinAdministratorsSid,
                null, pSid, cbSid);
//        System.out.println("Named pipe owner SID: " + getSIDString(ppSid));
//        System.out.println("Admin SID: " + getSIDString(pSid));
        //FIXME: wait till the pipe connection is released
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        return Advapi32.INSTANCE.EqualSid(pSid, ppSid);
    }

    /**
     * Create a sequence.
     *
     * @param driveId The drive ID.
     * @param authId  The authorization ID of the drive.
     */
    public void createSequence(String driveId, String authId) {
        Response res;
        try {
            String request = generateRequest(driveId, authId, RequestType.CreateSequence, null, null);
            write(request);
            String response = read();
            res = Response.Parse(response);
        } catch (Exception e) {
            throw new SequenceException("Could not create sequence: ", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SequenceException("Could not create sequence: " + res.error);
    }

    /**
     * Get a sequence by drive ID.
     *
     * @param driveId The drive ID.
     * @return The sequence.
     */
    public NonceSequence getSequence(String driveId) {
        Response res;
        try {
            String request = generateRequest(driveId, null, RequestType.GetSequence, null, null);
            write(request);
            String response = read();
            res = Response.Parse(response);
        } catch (Exception e) {
            throw new SequenceException("Could not get sequence", e);
        }
        if (res.status == Response.ResponseStatus.Error)
            throw new SequenceException("Could not get sequence: " + res.error);
        if (res.status == Response.ResponseStatus.NotFound)
            return null;
        return new NonceSequence(res.driveId, res.authId,
                res.nextNonce, res.maxNonce, res.seqStatus);
    }

    /**
     * Initialize the sequence.
     *
     * @param driveId    The drive ID.
     * @param authId     The auth ID of the device for the drive.
     * @param startNonce The starting nonce.
     * @param maxNonce   The maximum nonce.
     */
    public void initializeSequence(String driveId, String authId, byte[] startNonce, byte[] maxNonce) {
        Response res;
        try {
            String request = generateRequest(driveId, authId, RequestType.InitSequence,
                    startNonce, maxNonce);
            write(request);
            String response = read();
            res = Response.Parse(response);
        } catch (Exception e) {
            throw new SequenceException("Could not init sequence", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SequenceException("Could not init sequence: " + res.error);
    }

    /**
     * Get the next nonce
     *
     * @param driveId The drive ID.
     * @return The byte array with the next nonce.
     */
    public byte[] nextNonce(String driveId) {
        Response res;
        try {
            String request = generateRequest(driveId, null, RequestType.NextNonce, null, null);
            write(request);
            String response = read();
            res = Response.Parse(response);
        } catch (Exception e) {
            throw new SequenceException("Could not get next nonce", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SequenceException("Could not get next nonce: " + res.error);
        return res.nextNonce;
    }

    /**
     * Revoke a sequence by drive ID.
     *
     * @param driveId The drive Id
     */
    public void revokeSequence(String driveId) {
        Response res;
        try {
            String request = generateRequest(driveId, null, RequestType.RevokeSequence, null, null);
            write(request);
            String response = read();
            res = Response.Parse(response);
        } catch (Exception e) {
            throw new SequenceException("Could not revoke Sequence", e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SequenceException("Could not revoke Sequence: " + res.error);
    }

    /**
     * Set the max nonce this sequence can produce.
     *
     * @param driveId  The drive ID.
     * @param authId   The auth ID of the device for the drive.
     * @param maxNonce The maximum nonce.
     * @throws IOException If there was a problem with the sequencer.
     */
    public void setMaxNonce(String driveId, String authId, byte[] maxNonce) throws IOException {
        String request = generateRequest(driveId, authId, RequestType.SetMaxNonce,
                null, maxNonce);
        write(request);
        String response = read();
        Response res;
        try {
            res = Response.Parse(response);
        } catch (Exception e) {
            throw new SequenceException("Could not revoke Sequence: " + e);
        }
        if (res.status != Response.ResponseStatus.Ok)
            throw new SequenceException("Could not revoke Sequence: " + res.error);
    }

    private String generateRequest(String driveId, String authId, RequestType type,
                                   byte[] nextNonce, byte[] maxNonce)
            throws IOException {
        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        XMLStreamWriter out = null;
        try {
            XMLOutputFactory outputFactory = XMLOutputFactory.newInstance();
            out = outputFactory.createXMLStreamWriter(stream, "UTF-8");
            out.writeStartDocument("UTF-8", "1.0");
            out.writeStartElement("drive");
            out.writeAttribute("driveID", driveId);
            if (authId != null)
                out.writeAttribute("authID", authId);
            out.writeAttribute("type", type.toString());
            if (nextNonce != null)
                out.writeAttribute("nextNonce", Base64.getEncoder().encodeToString(nextNonce));
            if (maxNonce != null)
                out.writeAttribute("maxNonce", Base64.getEncoder().encodeToString(maxNonce));
            out.writeEndElement();
            out.writeEndDocument();
            out.writeCharacters(System.getProperty("line.separator"));
        } catch (XMLStreamException e) {
            throw new IOException(e);
        } finally {
            if (out != null) {
                try {
                    out.flush();
                    out.close();
                } catch (XMLStreamException e) {
                    throw new IOException("Could not close XML stream", e);
                }
            }
            stream.flush();
            stream.close();
        }
        return new String(stream.toByteArray());
    }

    /**
     * Close the sequencer.
     */
    public void close() {
        if (namedPipe != null) {
            try {
                namedPipe.close();
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    private static class Response {
        String driveId;
        String authId;
        ResponseStatus status;
        NonceSequence.Status seqStatus;
        byte[] nextNonce;
        byte[] maxNonce;
        String error;

        public enum ResponseStatus {
            /**
             * Success
             */
            Ok,
            /**
             * Not found
             */
            NotFound,
            /**
             * Error
             */
            Error
        }

        private static Response Parse(String contents) throws XPathExpressionException, ParserConfigurationException, IOException, SAXException {
            Response response = new Response();
            ByteArrayInputStream stream = new ByteArrayInputStream(contents.getBytes());
            try {
                DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
                DocumentBuilder builder = builderFactory.newDocumentBuilder();
                Document xmlDocument = builder.parse(stream);
                XPath xPath = XPathFactory.newInstance().newXPath();
                Node drive = (Node) xPath.compile("/drive").evaluate(xmlDocument, XPathConstants.NODE);
                if (drive != null) {
                    response.driveId = drive.getAttributes().getNamedItem("driveID").getNodeValue();
                    response.authId = drive.getAttributes().getNamedItem("authID").getNodeValue();
                    response.status = ResponseStatus.valueOf(drive.getAttributes().getNamedItem("status").getNodeValue());
                    if (drive.getAttributes().getNamedItem("nextNonce") != null) {
                        response.seqStatus = NonceSequence.Status.valueOf(drive.getAttributes().getNamedItem("seqStatus").getNodeValue());
                    }
                    if (drive.getAttributes().getNamedItem("nextNonce") != null) {
                        response.nextNonce = Base64.getDecoder().decode(drive.getAttributes().getNamedItem("nextNonce").getNodeValue());
                    }
                    if (drive.getAttributes().getNamedItem("maxNonce") != null) {
                        response.maxNonce = Base64.getDecoder().decode(drive.getAttributes().getNamedItem("maxNonce").getNodeValue());
                    }
                    if (drive.getAttributes().getNamedItem("error") != null) {
                        response.error = drive.getAttributes().getNamedItem("error").getNodeValue();
                    }
                }
            } catch (Exception ex) {
                ex.printStackTrace();
                throw ex;
            } finally {
                if (stream != null)
                    stream.close();
            }
            return response;
        }
    }

    private void write(String request) throws IOException {
        namedPipe.write(request.getBytes(StandardCharsets.UTF_8));
    }

    private String read() throws IOException {
        byte[] buffer = new byte[1024];
        int bytesRead = namedPipe.read(buffer, 0, buffer.length);
        return new String(buffer, 0, bytesRead);
    }

    private static String getSIDString(WinNT.PSID sid) {
        PointerByReference stringSid = new PointerByReference();
        Advapi32.INSTANCE.ConvertSidToStringSid(sid, stringSid);
        return stringSid.getValue().getWideString(0);
    }
}
